Add new export LLM config

jackzhxng opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #11162
* #11081
* #11029
* __->__ #11028

Differential Revision: [D75263991](https://our.internmc.facebook.com/intern/diff/D75263991)

jackzhxng committed:
Add new export LLM config

[ghstack-poisoned]

jackzhxng committed:
Update on "Add new export LLM config"




[ghstack-poisoned]

jackzhxng committed:
Update on "Add new export LLM config"




[ghstack-poisoned]

jackzhxng committed:
Update on "Add new export LLM config"




[ghstack-poisoned]

jackzhxng committed:
Update on "Add new export LLM config"




[ghstack-poisoned]

jackzhxng committed:
Update on "Add new export LLM config"




[ghstack-poisoned]

jackzhxng commented:
@jackzhxng has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D75263991).

jackzhxng commented:
@jackzhxng has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D75263991).

jackzhxng committed:
Update on "Add new export LLM config"


Differential Revision: [D75263991](https://our.internmc.facebook.com/intern/diff/D75263991)

[ghstack-poisoned]

jackzhxng commented:
@jackzhxng has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D75263991).

jackzhxng commented:
@jackzhxng has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D75263991).

jackzhxng committed:
Update on "Add new export LLM config"


Differential Revision: [D75263991](https://our.internmc.facebook.com/intern/diff/D75263991)

[ghstack-poisoned]

jackzhxng commented:
@jackzhxng has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D75263991).

jackzhxng commented:
@jackzhxng has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D75263991).

jackzhxng committed:
Update on "Add new export LLM config"


Differential Revision: [D75263991](https://our.internmc.facebook.com/intern/diff/D75263991)

[ghstack-poisoned]

jackzhxng committed:
Update on "Add new export LLM config"


Differential Revision: [D75263991](https://our.internmc.facebook.com/intern/diff/D75263991)

[ghstack-poisoned]

jackzhxng committed:
Update on "Add new export LLM config"


Differential Revision: [D75263991](https://our.internmc.facebook.com/intern/diff/D75263991)

[ghstack-poisoned]

jackzhxng committed:
Update on "Add new export LLM config"


Differential Revision: [D75263991](https://our.internmc.facebook.com/intern/diff/D75263991)

[ghstack-poisoned]

lucylq commented on examples/models/llama/config/llm_config.py:
```diff
@@ -0,0 +1,270 @@
+class BaseConfig:
+    """
+    These are specific to the specific model, e.g. whether it’s Qwen3 0.6B or Phi-4-mini.
+    For each of these different models, you can expect each of these fields to change.
```
Find this docstring a bit confusing.

> For each of these different models, you can expect each of these fields to change.

Do we expect variation in the fields for each model (ie, user/we add different fields), or the values?

Suggestion:
```
The BaseConfig sets the parameters required to construct the model in eager mode. Expect the values of each field to be different for each model (e.g. Qwen3 0.6B and Phi-4-mini will have different checkpoints).
```

lucylq commented on examples/models/llama/config/llm_config.py:
```diff
@@ -0,0 +1,270 @@
+################################################################################
+
+
+class ModelType(str, Enum):
```
Add a docstring here - could you add one for each class?

```
Describes the model family. Within each ModelType enum, there can be variation in the weight classes. For example, LLAMA3 comes in 1B, 3B and 8B. 
```

lucylq commented on examples/models/llama/config/llm_config.py:
```diff
@@ -0,0 +1,270 @@
+################################################################################
+
+
+class DtypeOverride(str, Enum):
```
Can we add docstring here as well, and also clarify when the model is converted? Iirc there's some logic to decide when to convert?

lucylq commented on examples/models/llama/config/llm_config.py:
```diff
@@ -0,0 +1,270 @@
+
+@dataclass
+class BackendConfig:
+    xnnpack: XNNPackConfig = field(default_factory=XNNPackConfig)
```
Wdyt about making these optional configs, set to None?

I think that could make it easier to tell when a backend is enabled or not, instead of looking through each backend config to check if enabled=False.

jackzhxng replied:
Mind if I follow up with this change? I would like to keep the changes in this diff as small as possible (ie as close to the original args as possible) to make it easier to pass all the internal CI

lucylq commented on examples/models/llama/config/llm_config.py:
```diff
@@ -0,0 +1,270 @@
+    """
+
+    dtype_override: DtypeOverride = DtypeOverride.FP32
+    enable_dynamic_shape: bool = True
```
Dynamic shapes default to true - does the user have to provide the dynamic shapes for export, or is it available for each ModelType?

jackzhxng replied:
Yeah, its the same for each model

lucylq replied:
Ok, asking because if a user brings their own model, we have to make sure that dynamic shapes exists, right? Maybe it's useful to add a comment specifying that.

lucylq replied:
ah, looks like it's provided by the builder, not on a per-model basis: https://github.com/pytorch/executorch/blob/main/extension/llm/export/builder.py#L186

jackzhxng committed:
Update on "Add new export LLM config"


Differential Revision: [D75263991](https://our.internmc.facebook.com/intern/diff/D75263991)

[ghstack-poisoned]

jackzhxng committed:
Update on "Add new export LLM config"


Differential Revision: [D75263991](https://our.internmc.facebook.com/intern/diff/D75263991)

[ghstack-poisoned]

jackzhxng committed:
Update on "Add new export LLM config"


Differential Revision: [D75263991](https://our.internmc.facebook.com/intern/diff/D75263991)

[ghstack-poisoned]

jackzhxng committed:
Update on "Add new export LLM config"


Differential Revision: [D75263991](https://our.internmc.facebook.com/intern/diff/D75263991)

[ghstack-poisoned]

jackzhxng committed:
Update on "Add new export LLM config"


Differential Revision: [D75263991](https://our.internmc.facebook.com/intern/diff/D75263991)

[ghstack-poisoned]

jackzhxng committed:
Update on "Add new export LLM config"


Differential Revision: [D75263991](https://our.internmc.facebook.com/intern/diff/D75263991)

[ghstack-poisoned]

jackzhxng committed:
Update on "Add new export LLM config"


Differential Revision: [D75263991](https://our.internmc.facebook.com/intern/diff/D75263991)

[ghstack-poisoned]

jackzhxng committed:
Update on "Add new export LLM config"


Differential Revision: [D75263991](https://our.internmc.facebook.com/intern/diff/D75263991)

[ghstack-poisoned]

kimishpatel commented on examples/models/llama/config/llm_config.py:
```diff
@@ -0,0 +1,495 @@
+        model_class: Which model to to export.
+        params: Model parameters, such as n_layers, hidden_size, etc.
+            If left empty will use defaults specified in model_args.py.
+        checkpoint: Path to the checkpoint file.
```
can this be hf path as well?

jackzhxng replied:
No, but at the moment if you specify a non-llama `model_class` and don't specify checkpoint, it will download from HF. Worth adding this comment

kimishpatel commented on examples/models/llama/config/llm_config.py:
```diff
@@ -0,0 +1,495 @@
+            e.g. '{"get_bos_id":128000, "get_eos_ids":[128009, 128001]}'
+        use_lora: Rank of the LoRA, if set to 0 then this means no LoRA. For use with QAT.
+        fairseq2: For legacy internal use cases, this is safe to ignore.
+        preq_mode: Legacy option to specify how prequantized weights are loaded.
+            Going forward, ExecuTorch supports loading weights prequantized through
+            TorchAo as-is, without any special handling.
+        preq_group_size: Legacy option to specify the group size of prequantized weights.
```
I thik we probably want to couple these together. If you are loading pre-quantized checkpoint that group size cannot be set independently, right? So maybe having a separate dataclass that captures all the params and maps it by name is better. ALthough I presume you have to keep this for BC?

kimishpatel commented on examples/models/llama/config/llm_config.py:
```diff
@@ -0,0 +1,495 @@
+@dataclass
+class ModelConfig:
+    """
+    Configurations not necessarily specific to the model, but are needed to
```
Then why call it modelconfig?

kimishpatel commented on examples/models/llama/config/llm_config.py:
```diff
@@ -0,0 +1,495 @@
+    and is source transform-based.
+    """
+
+    XNNPACK_DYNAMIC = "xnnpack_dynamic"
+    XNNPACK_DYNAMIC_QC4 = "xnnpack_dynamic_qc4"
```
we probably can remove these two

kimishpatel commented on examples/models/llama/config/llm_config.py:
```diff
@@ -0,0 +1,495 @@
+    extended_ops: bool = False
+
+
+class CoreMLQuantize(str, Enum):
+    B4W = "b4w"
+    C4W = "c4w"
```
maybe favor this pattern over the ones in pt2e_quantizer

kimishpatel commented on examples/models/llama/config/test_llm_config.py:
```diff
@@ -0,0 +1,104 @@
+class TestValidConstruction(unittest.TestCase):
+
+    def test_valid_llm_config(self):
+        LlmConfig(
```
are these configs constructible from json? I think that would e the best
